relay-sampling/src/evaluation.rs

```diff
@@ -7,226 +7,205 @@ use chrono::{DateTime, Utc};
 use rand::distributions::Uniform;
 use rand::Rng;
 use rand_pcg::Pcg32;
-use relay_base_schema::events::EventType;
-use relay_event_schema::protocol::Event;
+use relay_protocol::Getter;
 use serde::Serialize;
 use uuid::Uuid;
 
-use crate::config::{
-    DecayingFunction, RuleId, RuleType, SamplingConfig, SamplingMode, SamplingRule, SamplingValue,
-    TimeRange,
-};
-use crate::dsc::DynamicSamplingContext;
+use crate::config::{RuleId, SamplingRule, SamplingValue};
 
 /// Generates a pseudo random number by seeding the generator with the given id.
 ///
 /// The return is deterministic, always generates the same number from the same id.
-pub fn pseudo_random_from_uuid(id: Uuid) -> f64 {
+fn pseudo_random_from_uuid(id: Uuid) -> f64 {
     let big_seed = id.as_u128();
     let mut generator = Pcg32::new((big_seed >> 64) as u64, big_seed as u64);
     let dist = Uniform::new(0f64, 1f64);
     generator.sample(dist)
 }
 
-/// Returns an iterator of references that chains together and merges rules.
+/// State of the [`SamplingEvaluator`]'s rule matching.
 ///
-/// The chaining logic will take all the non-trace rules from the project and all the trace/unsupported
-/// rules from the root project and concatenate them.
-pub fn merge_rules_from_configs<'a>(
-    sampling_config: Option<&'a SamplingConfig>,
-    root_sampling_config: Option<&'a SamplingConfig>,
-) -> impl Iterator<Item = &'a SamplingRule> {
-    let transaction_rules = sampling_config
-        .into_iter()
-        .flat_map(|config| config.rules_v2.iter())
-        .filter(|&rule| rule.ty == RuleType::Transaction);
-
-    let trace_rules = root_sampling_config
-        .into_iter()
-        .flat_map(|config| config.rules_v2.iter())
-        .filter(|&rule| rule.ty == RuleType::Trace);
-
-    transaction_rules.chain(trace_rules)
+/// Enforces a pattern to avoid matching on more rules if a match has already been found.
+#[derive(Debug)]
+pub enum RuleMatchingState {
```

Comment:
jan: we discussed something like this in zadar. 

my thinking: 

1. It's now impossible to call `match_rules` if a match has already been found. 
2. We need the `rule_ids` in `SamplingMatch`, previous iteration we used std::mem::take which probably avoids allocation, however it leaves `SamplingEvaluator` in an invalid state. If we cloned it instead, we would have an unnecessary allocation. this implementaiton solves both issues, as we can do a partial move from `SamplingEvaluator` since it gets destroyed when the enum switches. 

inspect the `match_rules` and its callers to see what i mean

Reply:
_nit_: We can shorten the naming of this enum and its variants and introduce more semantics:

```rust
pub enum Evaluation {
    /// Matching has not completed and more rules can be evaluated.
    ///
    /// This can happen either if no active rules match the provided data, or if the matched rules
    /// are factors that require a final sampling value. In this case, the returned evaluator stores
    /// the state of the matched rules as well as the accumulated sampling factor and can be used to
    /// evaluate more rules.
    ///
    /// If evaluation returns `Continue` and there are no more rules to match, this should be
    /// considered as "no match".
    Continue(SamplingEvaluator),
    /// One or more rules have matched.
    Matched(SamplingMatch),
}
```

In a follow-up, we can try to use `std::ops::ControlFlow` for this. This implements `std::ops::Try`, which could allow us to write functions like this:

```rust
evaluator = evaluator.match_rules(seed, event, rules)?;
```

However, if I'm not mistaken that would require the containing function to also return a `ControlFlow`, which isn't exactly great.

Reply:
That looks a lot more clear, thx! 

and as mentioned I love the idea of using ControlFlow, will implement later!